add unit test coverage for pkg/util/node and Remove duplicate testcases for func TestGetNodeHostIPs

zhoumingcheng opened the pull request:
Signed-off-by: zhoumingcheng <zhoumingcheng@beyondcent.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Improve test coverage

Run test coverage command:
```
go test -cover ./pkg/util/node/
```
Befor:
```
ok      k8s.io/kubernetes/pkg/util/node/     0.019s  coverage: 60.3% of statements
```
After:
```
ok      k8s.io/kubernetes/pkg/util/node/   0.022s  coverage: 67.2% of statements
```

Remove duplicate testcases for func TestGetNodeHostIPs
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
Thank you！
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

zhoumingcheng committed:
add unit test coverage for pkg/util/node

Signed-off-by: zhoumingcheng <zhoumingcheng@beyondcent.com>

aojea commented on pkg/util/node/node_test.go:
```diff
@@ -259,3 +263,106 @@ func TestGetHostname(t *testing.T) {
 	}
 }
+
+func TestGetNodeIP(t *testing.T) {
```
this function is a wrapper on `GetNodeHostIP` , you can reuse the same testcases than `TestGetNodeHostIPs`

aojea commented on pkg/util/node/node_test.go:
```diff
@@ -259,3 +263,106 @@ func TestGetHostname(t *testing.T) {
+					{Type: v1.NodeExternalIP, Address: "d:c::b:a"},
+				}},
+			},
+			expect: []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 1, 2, 3, 4},
```
see `TestGetNodeHostIPs` for the format , this is hard to read

aojea commented on pkg/util/node/node_test.go:
```diff
@@ -259,3 +263,106 @@ func TestGetHostname(t *testing.T) {
+			expect: []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 1, 2, 3, 4},
+		},
+	}
+	for _, test := range testCases {
```
use `t.Run`

```
	for _, tc := range testcases {
		t.Run(tc.name, func(t *testing.T) {
```

aojea commented on pkg/util/node/node_test.go:
```diff
@@ -259,3 +263,106 @@ func TestGetHostname(t *testing.T) {
+			expect: false,
+		},
+	}
+	for _, test := range testCases {
```
https://github.com/kubernetes/kubernetes/pull/110825/files#r908187141

zhoumingcheng replied:
Thanks for your suggestion, I have changed all of the above. and removed duplicate testcases for func TestGetNodeHostIPs

aojea commented:
/ok-to-test
/sig-node

zhoumingcheng committed:
update unit test for pkg/util/node
Remove duplicate testcases for func TestGetNodeHostIPs

Signed-off-by: zhoumingcheng <zhoumingcheng@beyondcent.com>

zhoumingcheng commented:
/retest

zhoumingcheng commented:
/assign @thockin
please take a look, thanks!

thockin commented on pkg/util/node/node_test.go:
```diff
@@ -259,3 +231,201 @@ func TestGetHostname(t *testing.T) {
+					t.Error("create node error")
+				}
+			}
+			result := GetNodeIP(kubeClientSet, test.nodeName)
```
Tell me if I am misunderstanding:

Testing through `GetNodeIP()` seems to give coverage of the `wait.ExponentialBackoff()` logic but ultimately calls `GetNodeHostIP()`.  This test doesn't actually exercise any of the retry logic, so you're moving test cases around to score coverage "points" but it's bogus coverage.

I think I'd rather see all the test cases in the ` GetNodeHostIPs()` (plural) test, with a subset of those in a test for `GetNodeHostIP()` (singular).  This function (`GetNodeIP()`) seems to only be called from 2 places and I think it would be better to move this function to be internal to both.  "A little copy is better than a little dep".

zhoumingcheng replied:
Thanks for the suggestion, I did notice that GetNodeIP() ends up calling GetNodeHostIP(). My GetNodeIP() test case and TestGetNodeHostIPs() both cover GetNodeHostIP(). I wonder if it's possible to cover wait.ExponentialBackoff() over GetNodeHostIP() directly via TestGetNodeIP(). So I removed TestGetNodeHostIPs. Not sure if this is possible?

thockin replied:
I think this moves in the wrong direction.  Tests should focus on the lowest level possible.

I'd set this up as a comprehensive test of `GetNodeHostIPs()` - since that seems to be where the real work is happening.

Testing `GetNodeIP()` (if done at all) should focus on the logic that is unique to that function, which is the backoff (which is mostly boilerplate and does not really belong in this utils lib in the first place).

zhoumingcheng replied:
I'm sorry, with all your replies to me, I'm a little bit confused about what you mean here. Do I wish I could move the tests for GetNodeIP() to the called function directory, or should the tests be concentrated at the lowest level possible. If it is the second, I will remove the test cases that are repeated with TestGetNodeHostIPs() in TestGetNodeIP(). Then TestGetNodeIP() will only focus on testing GetNodeIP(). Not sure if I misunderstood what you meant if I changed it like this. Or should I move the test of GetNodeIP() to the corresponding test function of the called function, such as pkg/util/node/node_test.go Test_detectNodeIP(). Sorry to bother you.

thockin commented:
Don't bother to test GetNodeIP() until the lower level functions have
complete, full coverage.

On Wed, Jul 6, 2022, 7:03 PM zhoumingcheng ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/util/node/node_test.go
> <https://github.com/kubernetes/kubernetes/pull/110825#discussion_r915384570>
> :
>
> > +					{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
> +					{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
> +				}},
> +			},
> +			expect: netutils.ParseIPSloppy("a:b::c:d"),
> +		},
> +	}
> +	for _, test := range testCases {
> +		t.Run(test.name, func(t *testing.T) {
> +			kubeClientSet := kubeclientfake.NewSimpleClientset()
> +			if test.node != nil {
> +				if _, err := kubeClientSet.CoreV1().Nodes().Create(context.Background(), test.node, metav1.CreateOptions{}); err != nil {
> +					t.Error("create node error")
> +				}
> +			}
> +			result := GetNodeIP(kubeClientSet, test.nodeName)
>
> I'm sorry, with all your replies to me, I'm a little bit confused about
> what you mean here. Do I wish I could move the tests for GetNodeIP() to the
> called function directory, or should the tests be concentrated at the
> lowest level possible. If it is the second, I will remove the test cases
> that are repeated with TestGetNodeHostIPs() in TestGetNodeIP(). Then
> TestGetNodeIP() will only focus on testing GetNodeIP(). Not sure if I
> misunderstood what you meant if I changed it like this. Or should I move
> the test of GetNodeIP() to the corresponding test function of the called
> function, such as pkg/util/node/node_test.go Test_detectNodeIP(). Sorry to
> bother you.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/110825#discussion_r915384570>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVASYAJFSJKLSZJI35LVSY3HFANCNFSM52BHMHRQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

zhoumingcheng committed:
update test func for pkg/util/node

Signed-off-by: zhoumingcheng <zhoumingcheng@beyondcent.com>

zhoumingcheng commented:
/retest

zhoumingcheng commented:
> Don't bother to test GetNodeIP() until the lower level functions have complete, full coverage.
> […](#)
> On Wed, Jul 6, 2022, 7:03 PM zhoumingcheng ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In pkg/util/node/node_test.go <[#110825 (comment)](https://github.com/kubernetes/kubernetes/pull/110825#discussion_r915384570)> : > + {Type: v1.NodeExternalIP, Address: "4.3.2.1"}, + {Type: v1.NodeExternalIP, Address: "4.3.2.2"}, + }}, + }, + expect: netutils.ParseIPSloppy("a🅱️:c:d"), + }, + } + for _, test := range testCases { + t.Run(test.name, func(t *testing.T) { + kubeClientSet := kubeclientfake.NewSimpleClientset() + if test.node != nil { + if _, err := kubeClientSet.CoreV1().Nodes().Create(context.Background(), test.node, metav1.CreateOptions{}); err != nil { + t.Error("create node error") + } + } + result := GetNodeIP(kubeClientSet, test.nodeName) I'm sorry, with all your replies to me, I'm a little bit confused about what you mean here. Do I wish I could move the tests for GetNodeIP() to the called function directory, or should the tests be concentrated at the lowest level possible. If it is the second, I will remove the test cases that are repeated with TestGetNodeHostIPs() in TestGetNodeIP(). Then TestGetNodeIP() will only focus on testing GetNodeIP(). Not sure if I misunderstood what you meant if I changed it like this. Or should I move the test of GetNodeIP() to the corresponding test function of the called function, such as pkg/util/node/node_test.go Test_detectNodeIP(). Sorry to bother you. — Reply to this email directly, view it on GitHub <[#110825 (comment)](https://github.com/kubernetes/kubernetes/pull/110825#discussion_r915384570)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABKWAVASYAJFSJKLSZJI35LVSY3HFANCNFSM52BHMHRQ> . You are receiving this because you were assigned.Message ID: ***@***.***>

Thanks for your suggestion, I have removed TestGetNodeIP() and removed the duplicated test cases in TestGetNodeHostIPs().

thockin commented:
Thanks!

/lgtm
/approve

pacoxu commented:
/sig node

zhoumingcheng commented:
/retest
